powerauth-nextstep/src/main/java/io/getlime/security/powerauth/app/nextstep/service/CredentialService.java

```diff
@@ -624,9 +631,14 @@ public CredentialSecretDetail createCredential(UserIdentityEntity user, Credenti
         credential.setAttemptCounter(0);
         credential.setFailedAttemptCounterSoft(0);
         credential.setFailedAttemptCounterHard(0);
-        credentialRepository.save(credential);
+        if (newCredential) {
+            // Credential needs to be added after all validations, otherwise JPA may save the credential prematurely
+            user.getCredentials().add(credential);
+        }
+
         // Save credential into credential history
-        credentialHistoryService.createCredentialHistory(credential, new Date());
+        credentialHistoryService.createCredentialHistory(user, credential, new Date());
+        userIdentityRepository.save(user);
```

Comment:
Should you assign to `user` here?

```java
user = userIdentityRepository.save(user);
```

You are working with the `user` variable later in the code...

Reply:
When I do this, IntelliJ IDEA marks the variable as unused, so probably not.

Reply:
You do use the variable on line 688, se below. This was not the case before (and so the variable was unused), but now it is the case...

Reply:
Ignore... I see this, there was one more collapse in the code review block... My bad.